server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -317,14 +328,112 @@ public IndexInput openInput(String name, IOContext context) throws IOException {
         }
     }
 
-    public void copyFrom(Directory from, String src, String dest, IOContext context, boolean useCommonSuffix) throws IOException {
+    /**
+     * Copies a list of files from the source directory to a remote based on multi-stream upload support.
+     * If vendor plugin supports uploading multiple parts in parallel, <code>BlobContainer#writeBlobByStreams</code>
+     * will be used, else, the legacy {@link RemoteSegmentStoreDirectory#copyFrom(Directory, String, String, IOContext, boolean)}
+     * will be called.
+     *
+     * @param from    The directory for all files to be uploaded
+     * @param files   A list containing the names of all files to be uploaded
+     * @param context IOContext to be used to open IndexInput to files during remote upload
+     * @throws Exception When upload future creation fails or if {@link RemoteSegmentStoreDirectory#copyFrom(Directory, String, String, IOContext, boolean)}
+     *                   throws an exception
+     */
+    public void copyFilesFrom(Directory from, Collection<String> files, IOContext context) throws Exception {
+
+        List<CompletableFuture<UploadResponse>> resultFutures = new ArrayList<>();
+
+        for (String src : files) {
+            String remoteFilename = createRemoteFileName(src, false);
+            if (remoteDataDirectory.getBlobContainer().isMultiStreamUploadSupported()) {
+                CompletableFuture<UploadResponse> resultFuture = createUploadFuture(from, src, remoteFilename, context);
+                resultFutures.add(resultFuture);
```

Comment:
I think we should be able to move some of this logic for parallel upload to one level higher and use it with existing copyFrom method itself because each upload itself is independent from the rest.

Reply:
One level higher is `RemoteStoreRefreshListener` which doesn't seem as the right place to put upload code. Secondly, each upload is submitted to a batch of upload. Current abstractions do not allow parallel upload of files. Each upload is logically independent of each other but current abstractions defined in directory structure are not compatible with our use case. If we want to really treat each upload separately then either we have to rewrite directory abstractions to support async transfers or use the existing approach of batch upload.
We cannot compromise on performance for the sake of maintenance.